clients/include/rocblas_arguments.hpp

```diff
@@ -84,91 +87,89 @@ struct Arguments
     rocblas_initialization initialization;
     char                   known_bug_platforms[64];
 
-    // Validate input format.
-    // rocblas_gentest.py is expected to conform to this format.
-    // rocblas_gentest.py uses rocblas_common.yaml to generate this format.
-    static void validate(std::istream& ifs)
+    /*************************************************************************
+     *                     End Of Arguments                                  *
+     *************************************************************************/
+
+    // clang-format off
+
+// Generic macro which operates over the list of arguments in order of declaration
+#define FOR_EACH_ARGUMENT(OPER, SEP) \
+    OPER(M) SEP                      \
+    OPER(N) SEP                      \
+    OPER(K) SEP                      \
+    OPER(KL) SEP                     \
+    OPER(KU) SEP                     \
+    OPER(lda) SEP                    \
+    OPER(ldb) SEP                    \
+    OPER(ldc) SEP                    \
+    OPER(ldd) SEP                    \
+    OPER(a_type) SEP                 \
+    OPER(b_type) SEP                 \
+    OPER(c_type) SEP                 \
+    OPER(d_type) SEP                 \
+    OPER(compute_type) SEP           \
+    OPER(incx) SEP                   \
+    OPER(incy) SEP                   \
+    OPER(incd) SEP                   \
+    OPER(incb) SEP                   \
+    OPER(alpha) SEP                  \
+    OPER(alphai) SEP                 \
+    OPER(beta) SEP                   \
+    OPER(betai) SEP                  \
+    OPER(transA) SEP                 \
+    OPER(transB) SEP                 \
+    OPER(side) SEP                   \
+    OPER(uplo) SEP                   \
+    OPER(diag) SEP                   \
+    OPER(batch_count) SEP            \
+    OPER(stride_a) SEP               \
+    OPER(stride_b) SEP               \
+    OPER(stride_c) SEP               \
+    OPER(stride_d) SEP               \
+    OPER(stride_x) SEP               \
+    OPER(stride_y) SEP               \
+    OPER(norm_check) SEP             \
+    OPER(unit_check) SEP             \
+    OPER(timing) SEP                 \
+    OPER(iters) SEP                  \
+    OPER(algo) SEP                   \
+    OPER(solution_index) SEP         \
+    OPER(flags) SEP                  \
+    OPER(function) SEP               \
+    OPER(name) SEP                   \
+    OPER(category) SEP               \
+    OPER(initialization) SEP         \
+    OPER(known_bug_platforms)
+
+#define COMMA_SEPARATOR ,
+
+    // clang-format on
+
+    /**************************************************************
+    * Tuple of argument name, value pairs, preserving references *
+    **************************************************************/
+#define NAME_VALUE_PAIR(NAME) #NAME, NAME
+    auto as_tuple() const
```

Comment:
If the point of the tuple is to just to pack and then unpack, it could be simpler(and probably faster to compile) to use a lambda instead that visits each pair:

```cpp
#define NAME_VALUE_PAIR(NAME) f(#NAME, NAME);
auto visit_each() {
    return [=](auto f) {
        FOR_EACH_ARGUMENT(NAME_VALUE_PAIR,)
    };
}
```

And then printing could be implemented like this:

```cpp
template <typename Visitor>
static rocblas_ostream& print_pairs(rocblas_ostream& os, const Visitor& v)
{
    // Turn YAML formatting on
    os << rocblas_ostream::yaml_on;

    const char* delim = "{ ";
    v([&](const char* name, const auto& value) {
        os << delim << name << ": " << value;
        delim = ", ";
    });

    // Closing brace and turn YAML formatting off
    return os << " }\n" << rocblas_ostream::yaml_off;
}
```

And there is no need for dealing with indices or tuples.

Reply:
No, besides packing/unpacking, the point of using `std::tuple` is also to perform associative lookup (e.g., `std::unordered_map`). We need to track every argument regardless of type, i.e., to perform associative container operations in a type-generic way.

Reply:
`std::tuple` is a heterogeneous literal type which can be `hash`ed (although we write our own `hash` and `less` functions).

Almost all of the code you're talking about is in our testing code, not in our main library code. The `Arguments` class is specifically for our testing code; it's a universal way of expressing all arguments and external parameters of a particular problem to be tested. It is supposed to be cross-language (C/C++/Python/YAML), which is why it's so POD-ish.

We want the methods applied to `Arguments` to be simple and universal, to make it easier to write tests for dozens of different routines, rather than having the most efficient compile-time or run-time.

Reply:
For the printing of `(name, value)` pairs, a tuples implementation has been replaced with C macro iteration (`FOR_EACH_ARGUMENT`). But for the tracking and recording of `Arguments`, `std::tuple` is still necessary because it's stored in a `std::unordered_map` as a key.

The limit is not reached unless a large percentage of  `Arguments` are used to construct a `std::tuple`. With a subset of `Arguments` for a given problem, the limit is not reached. As stated earlier, this is due to a bad `libstdc++` implementation of `std::is_nothrow_move_constructible`.